Paper: Using Numba for GPU acceleration of Neutron Beamline Digital Twins

ckendrick opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@mepa<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@matthewfeickert, @sanhitamj<!--end-reviewers-list-->

ckendrick committed:
Initial version of paper

deniederhut commented:
@scoobies assign @mepa as editor

scoobies commented:
Assigned! @mepa is now the editor

mepa commented:
@scoobies build status

scoobies commented:
{"ckendrick-2023_mcvine"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpeo_7_np7'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: coleman_kendrick\n*** Warning: /tmp/tmpeo_7_np7/publisher/../output/coleman_kendrick/paper_stats.json does not exist.\n*** Warning: /tmp/tmpeo_7_np7/publisher/_build/toc.json does not exist.\n*** Warning: /tmp/tmpeo_7_np7/publisher/_build/other.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/ckendrick-2023_mcvine.pdf", "build_status"=>"success", "build_timestamp"=>"02/06 17:32"}, "status"=>"success"}}

mepa commented:
@scoobies check references

scoobies commented:
Checking the BibTeX entries failed with the following error: 
```
Couldn't find bibliography entry
```

mepa commented:
@scoobies check references

scoobies commented:
Checking the BibTeX entries failed with the following error: 
```
Couldn't find bibliography entry
```

mepa commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.3233/JNR-190108 is OK
- 10.3233/JNR-200186 is OK
- 10.1088/2399-6528/ab3622 is OK
- 10.1103/PhysRevB.107.094511 is OK
- 10.13139/ORNLNCCS/1975747 is OK

MISSING DOIs

- 10.1016/j.physb.2018.11.042 may be a valid DOI for title: Design of a radial collimator for the SEQUOIA direct geometry chopper spectrometer
- 10.1063/1.5007089 may be a valid DOI for title: Design and operating characteristic of a vacuum furnace for time-of-flight inelastic neutron scattering measurements
- 10.1063/1.5116147 may be a valid DOI for title: Super-resolution energy spectra from neutron direct-geometry spectrometers
- 10.1063/5.0079031 may be a valid DOI for title: A super-resolution technique to analyze single-crystal inelastic neutron scattering measurements using direct-geometry chopper spectrometers
- 10.1103/physrevb.105.l180403 may be a valid DOI for title: Damped Dirac magnon in the metallic kagome antiferromagnet FeSn
- 10.1063/5.0086451 may be a valid DOI for title: A concept of a broadband inverted geometry spectrometer for the Second Target Station at the Spallation Neutron Source
- 10.1063/5.0089783 may be a valid DOI for title: MENUS—Materials engineering by neutron scattering
- 10.1063/5.0089740 may be a valid DOI for title: CHESS: The future direct geometry spectrometer at the second target station
- 10.1063/5.0090919 may be a valid DOI for title: VERDI: VERsatile DIffractometer with wide-angle polarization analysis for magnetic structure studies in powders and single crystals
- 10.1063/5.0090810 may be a valid DOI for title: EWALD: A macromolecular diffractometer for the second target station
- 10.1063/5.0089524 may be a valid DOI for title: PIONEER, a high-resolution single-crystal polarized neutron diffractometer
- 10.1063/5.0090527 may be a valid DOI for title: CENTAUR—The small-and wide-angle neutron scattering diffractometer/spectrometer for the Second Target Station of the Spallation Neutron Source
- 10.1063/5.0089349 may be a valid DOI for title: EXPANSE: A time-of-flight expanded angle neutron spin echo spectrometer at the Second Target Station of the Spallation Neutron Source
- 10.1063/5.0122279 may be a valid DOI for title: Cinematic reflectometry using QIKR, the quite intense kinetics reflectometer
- 10.1016/j.nima.2022.167881 may be a valid DOI for title: A realistic guide misalignment model for the Second Target Station instruments at the Spallation Neutron Source
- 10.1063/1.4870050 may be a valid DOI for title: A comparison of four direct geometry time-of-flight spectrometers at the Spallation Neutron Source
- 10.1088/1742-6596/251/1/012058 may be a valid DOI for title: SEQUOIA: A newly operating chopper spectrometer at the SNS
- 10.1038/ncomms2117 may be a valid DOI for title: Quantum oscillations of nitrogen atoms in uranium nitride
- 10.1016/j.nima.2015.11.118 may be a valid DOI for title: MCViNE–An object oriented Monte Carlo neutron ray tracing simulation package
- 10.1103/physrevb.89.144302 may be a valid DOI for title: Using Monte Carlo ray tracing simulations to model the quantum harmonic oscillator modes observed in uranium nitride

INVALID DOIs

- https://doi.org/10.1016/j.cpc.2017.04.012 is INVALID because of 'https://doi.org/' prefix
```

deniederhut commented:
@scoobies add @matthewfeickert to reviewers

scoobies commented:
@matthewfeickert added to the reviewers list!

deniederhut commented:
@scoobies code of conduct

scoobies commented:
Our CoC: https://www.scipy2023.scipy.org/code-of-conduct

scoobies commented:
Reports of abusive or harassing behavior may be reported to scipy@enthought.com

ckendrick committed:
Fix invalid DOI

ckendrick commented:
@mepa I just fixed the invalid DOI from the reference check report above. Let me know if there is anything else I need to do.

matthewfeickert commented:
@scoobies help

scoobies commented:
Hello @matthewfeickert, here are the things you can ask me to do:

```

# List all available commands
@scoobies help

# Show our community Code of Conduct and Guidelines
@scoobies code of conduct

# Builds paper
@scoobies build paper

# Checks build status
@scoobies build status

# Check the references of the paper for missing DOIs
# Optionally, it can be run on a non-default branch 
@scoobies check references
@scoobies check references from branch custom-branch-name

# Label issue with: paper
@scoobies mark paper

# Label issue with: does_not_build:server
@scoobies mark server fail

# Remove labels: does_not_build:server
@scoobies mark server success

```

matthewfeickert commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.3233/JNR-190108 is OK
- 10.3233/JNR-200186 is OK
- 10.1088/2399-6528/ab3622 is OK
- 10.1103/PhysRevB.107.094511 is OK
- 10.13139/ORNLNCCS/1975747 is OK
- 10.1016/j.cpc.2017.04.012 is OK

MISSING DOIs

- 10.1016/j.physb.2018.11.042 may be a valid DOI for title: Design of a radial collimator for the SEQUOIA direct geometry chopper spectrometer
- 10.1063/1.5007089 may be a valid DOI for title: Design and operating characteristic of a vacuum furnace for time-of-flight inelastic neutron scattering measurements
- 10.1063/1.5116147 may be a valid DOI for title: Super-resolution energy spectra from neutron direct-geometry spectrometers
- 10.1063/5.0079031 may be a valid DOI for title: A super-resolution technique to analyze single-crystal inelastic neutron scattering measurements using direct-geometry chopper spectrometers
- 10.1103/physrevb.105.l180403 may be a valid DOI for title: Damped Dirac magnon in the metallic kagome antiferromagnet FeSn
- 10.1063/5.0086451 may be a valid DOI for title: A concept of a broadband inverted geometry spectrometer for the Second Target Station at the Spallation Neutron Source
- 10.1063/5.0089783 may be a valid DOI for title: MENUS—Materials engineering by neutron scattering
- 10.1063/5.0089740 may be a valid DOI for title: CHESS: The future direct geometry spectrometer at the second target station
- 10.1063/5.0090919 may be a valid DOI for title: VERDI: VERsatile DIffractometer with wide-angle polarization analysis for magnetic structure studies in powders and single crystals
- 10.1063/5.0090810 may be a valid DOI for title: EWALD: A macromolecular diffractometer for the second target station
- 10.1063/5.0089524 may be a valid DOI for title: PIONEER, a high-resolution single-crystal polarized neutron diffractometer
- 10.1063/5.0090527 may be a valid DOI for title: CENTAUR—The small-and wide-angle neutron scattering diffractometer/spectrometer for the Second Target Station of the Spallation Neutron Source
- 10.1063/5.0089349 may be a valid DOI for title: EXPANSE: A time-of-flight expanded angle neutron spin echo spectrometer at the Second Target Station of the Spallation Neutron Source
- 10.1063/5.0122279 may be a valid DOI for title: Cinematic reflectometry using QIKR, the quite intense kinetics reflectometer
- 10.1016/j.nima.2022.167881 may be a valid DOI for title: A realistic guide misalignment model for the Second Target Station instruments at the Spallation Neutron Source
- 10.1063/1.4870050 may be a valid DOI for title: A comparison of four direct geometry time-of-flight spectrometers at the Spallation Neutron Source
- 10.1088/1742-6596/251/1/012058 may be a valid DOI for title: SEQUOIA: A newly operating chopper spectrometer at the SNS
- 10.1038/ncomms2117 may be a valid DOI for title: Quantum oscillations of nitrogen atoms in uranium nitride
- 10.1016/j.nima.2015.11.118 may be a valid DOI for title: MCViNE–An object oriented Monte Carlo neutron ray tracing simulation package
- 10.1103/physrevb.89.144302 may be a valid DOI for title: Using Monte Carlo ray tracing simulations to model the quantum harmonic oscillator modes observed in uranium nitride

INVALID DOIs

- None
```

matthewfeickert commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper ckendrick-2023_mcvine scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

matthewfeickert commented:
Note to self: While the build status and current builds of all papers are linked on https://procbuild.scipy.org/ the current build for this paper is available at https://procbuild.scipy.org/download/ckendrick-2023_mcvine

deniederhut commented:
@scoobies add @sanhitamj to reviewers

scoobies commented:
@sanhitamj added to the reviewers list!

sanhitamj commented on papers/coleman_kendrick/metadata.yaml:
```diff
@@ -0,0 +1,24 @@
+abstract:
+- Digital twins of neutron instruments using Monte Carlo ray tracing have proven to be useful in neutron data analysis and verifying instrument and sample designs. However, these simulations can become quite complex and computationally demanding with tens of billions of neutrons. In this paper, we present a GPU accelerated version of MCViNE using Python and Numba to balance user extensibility wit ...
+keywords: 
+  - monte carlo
```
Would it be better to say 'Monte Carlo', first letters in uppercase?

sanhitamj commented on papers/coleman_kendrick/metadata.yaml:
```diff
@@ -0,0 +1,24 @@
+abstract:
+- Digital twins of neutron instruments using Monte Carlo ray tracing have proven to be useful in neutron data analysis and verifying instrument and sample designs. However, these simulations can become quite complex and computationally demanding with tens of billions of neutrons. In this paper, we present a GPU accelerated version of MCViNE using Python and Numba to balance user extensibility wit ...
```
Some of these acronyms would be new for a general SciPy attendee. It'd be useful to have those spelled - `JIT`, `LLVM`.

sanhitamj commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+\section{Introduction}
+
+MCViNE\cite{lin2016mcvine,Lin_2019} is a software package for creating digital twins of neutron scattering experiments using a Monte Carlo ray-tracing approach. 
+In this method, randomly generated probability packets (representing neutrons) are propagated through a series of components. Each component changes the probability packets according to the physics of the component. 
```
To understand 'physics for the component' needs some explanation for what is a component, or perhaps an example?

Since better explanation for what is a component follows - 'type of the component'?

sanhitamj commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+These simulations are useful in performing advanced neutron data analysis \cite{lin2022super,islam2019super,sala2022ferrimagnetic,do2022damped,leiner2019frustrated,lin2014UN} as well as in designing novel neutron instruments\,\cite{mamontov2022bwaves,an2022menus,sala2022chess,garlea2022verdi,leighton2022ewald,liu2022pioneer,qian2022centaur,changwoo2022expanse,ankner2023cinematic,hassina2023cupi2d ...
+Currently, MCViNE only runs on CPUs which is a bottleneck in large simulations with tens of billions of neutrons, and in modelling complex multiple scattering, with some simulations taking months to complete. Due to the massively parallel nature of Monte Carlo methods, bringing GPU acceleration to these simulations would offer superior performance and scalability. MCViNE was originally implemente ...
+
+To further improve performance and to create an easily extensible code base, Python and Numba\cite{lam2015numba} were chosen to create a new package of MCViNE providing GPU acceleration, \texttt{mcvine.acc}\footnote{\accrepo}. Numba is an open-source JIT (just-in-time) compiler for Python using LLVM to generate efficient machine code and supports GPUs using NVIDIA CUDA. Numba is designed for scie ...
```
1. Glad to see the full form of `JIT`. The same for `LLVM` as well?
2. Someone who hasn't used Numba would get more curious with - "we are only using Numba for its GPU capabilities" - as to what the rest of the capabilities are, and why aren't those used. 
"Only" is begging to ask more questions.

matthewfeickert replied:
```suggestion
To further improve performance and to create an easily extensible code base, Python and Numba\cite{lam2015numba} were chosen to create a new package of MCViNE providing GPU acceleration, \texttt{mcvine.acc}\footnote{\accrepo}. Numba is an open-source JIT (just-in-time) compiler for Python using LLVM to generate efficient machine code and supports GPUs using NVIDIA CUDA. Numba is designed for scientific computing and can support NumPy arrays and functions. Currently, we are only using Numba for its GPU capabilities. This accelerated MCViNE package is compatible with existing MCViNE scripts, and using a mixture of CPU and GPU components is supported.
```

`\texttt{mcvine.acc}\footnote{\accrepo}` Don't make https://github.com/mcvine/acc a footnote. This code deserves to be a full citation (consider adding a [CITATION.cff](https://citation-file-format.github.io/) to your repo :rocket:). Is this code going to be/or is currently distributed to PyPI or conda-forge?

sanhitamj commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+\begin{figure}[h]
+    \centering
+    \includegraphics[width=\columnwidth]{figures/mcvine_hss_isotropic_timings.png}
+    \caption{Time versus neutron counts for CPU (blue line) and GPU (orange line) for a simple instrument using a homogeneous single scatterer (HSS) with a aluminum sphere sample.}
```
Green line is multi-core CPU. Perhaps that mention will be useful here in the caption as well. It may be worth mentioning explicitly how many cores of the GPU - either in the plot or in the caption or both.

Same comment for the next 2 figure captions as well.

matthewfeickert replied:
+1 to this please.

sanhitamj commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+The N vibrates as a harmonic oscillator which provides equally spaced lines in energy transfer $E$.
+The lines are modeled in MCViNE with a sample scattering kernel containing a composite of $E(Q)$ kernels with constant $E$ values of a 50 meV spacing from 0 to 350 meV.
+
+Figure \ref{fig:results_isotropic_sphere} shows the performance of the isotropic sphere sample for the CPU version of MCViNE with one and 16 cores (blue and green curves) compared to the GPU version of MCViNE (orange curve). At $10^{12}$ neutrons, \texttt{mcvine.acc} achieves a speedup of 1725x over a single core and 133x over 16 cores.
```
It is redundant to add here that this figure is - for a simple instrument using a homogeneous single scatterer (HSS) with a aluminum sphere sample - but may be useful for the uninitiated.

(I had to go a bit back and forth to make sure all 3 figures are different.)

matthewfeickert replied:
Improved figure captions that do more than state what quantities are plotted would help here.

sanhitamj commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+    \label{fig:results_ms}
+\end{figure}
+
+For the UN single scattering case, figure \ref{fig:results_ss} shows that for $10^{11}$ neutrons, the GPU version obtained a speedup of 383x over a single core, and 33x over 16 cores. For the UN multiple scattering case, figure \ref{fig:results_ms} shows that for $10^{10}$ neutrons, the GPU version obtained a speedup of 137x over a single core, and 10x over 16 cores. Comparing the speedup achieve ...
```
A table format to display the speed up for various samples?

matthewfeickert replied:
+1 to this please.

sanhitamj commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+\label{fig:UN_plot}
+\end{figure}
+
+In this case the incident beamline simulation (the simulation up to the sample containing the SNS source, guide choppers and slits) for the ARCS instrument\cite{stone2014comparison} was run using McStas\cite{Willendrup:2020aa,Willendrup:2021aa} inside a workflow tool\cite{10.1007/978-3-031-23606-8_9}.
```
Full forms for `SNS` and `ARCS` please.

sanhitamj commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+For the first study, we focus on the performance gain achieved by the GPU accelerated version of MCViNE.
+We used a simple instrument consisting of a source, sample, and detector to focus on the sample assembly component. 
+We performed tests with two different samples: a simple spherical sample with an isotropic scattering kernel, and a second with a more complex Uranium Nitride (UN) sample.
```
It would be useful for a reader to know why Aluminium sphere and UN samples were used. Say, for example, that is the standard, or something short.

sanhitamj reviewed (changes requested):
Well written; a few minor comments for better readability.

matthewfeickert commented on papers/coleman_kendrick/bib.bib:
```diff
@@ -0,0 +1,384 @@
+%% This BibTeX bibliography file was created using BibDesk.
```
Most of these references are missing DOIs and/or URLs. This needs to be corrected.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+\section{Introduction}
+
+MCViNE\cite{lin2016mcvine,Lin_2019} is a software package for creating digital twins of neutron scattering experiments using a Monte Carlo ray-tracing approach. 
```
@mepa is it possible to have the template style show these as "MCViNE[1,2]" instead of "MCViNE[1], [2]"? This looks pretty bad later on when there are multiple citations that would look much nicer as "[3-8]" instead of "[3], [4], [5], [6], [7], [8]".

mepa replied:
@matthewfeickert, the Proceedings use IEEE style citations (see https://pitt.libguides.com/citationhelp/ieee, for example) so each source should have its own brackets. The above does need a space between `MCViNE` and the brackets, though, i.e., `MCViNE \cite{lin2016mcvine,Lin_2019}`.

For multiple citations, one could render "[3] - [8]", if preferred, by citing the first and last sources in the series, separated by a dash, e.g., `\cite{source3} - \cite{source8}`.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+In this method, randomly generated probability packets (representing neutrons) are propagated through a series of components. Each component changes the probability packets according to the physics of the component. 
+There is no correlation between packets, so the system is embarrassingly parallel.
+These simulations are useful in performing advanced neutron data analysis \cite{lin2022super,islam2019super,sala2022ferrimagnetic,do2022damped,leiner2019frustrated,lin2014UN} as well as in designing novel neutron instruments\,\cite{mamontov2022bwaves,an2022menus,sala2022chess,garlea2022verdi,leighton2022ewald,liu2022pioneer,qian2022centaur,changwoo2022expanse,ankner2023cinematic,hassina2023cupi2d ...
+Currently, MCViNE only runs on CPUs which is a bottleneck in large simulations with tens of billions of neutrons, and in modelling complex multiple scattering, with some simulations taking months to complete. Due to the massively parallel nature of Monte Carlo methods, bringing GPU acceleration to these simulations would offer superior performance and scalability. MCViNE was originally implemente ...
```
```suggestion
Currently, MCViNE only runs on CPUs which is a bottleneck in large simulations with tens of billions of neutrons, and in modelling complex multiple scattering, with some simulations taking months to complete. Due to the massively parallel nature of Monte Carlo methods, bringing GPU acceleration to these simulations would offer superior performance and scalability. MCViNE was originally implemented in C++ and parallelized using MPI, with bindings to Python for user interaction.
However, extensibility for the end user can be very difficult.
```

Improper use of a semicolon, use two sentences instead.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+\subsection{MCViNE Overview}
+
+MCViNE simulations are run from a script that defines an instrument, where an instrument is composed of multiple components. A simple 4 component instrument is shown in figure \ref{fig:instrument}. Full instruments may have several hundred components. Each instrument script is run with a specified number of probability packets, where each packet has several state variables: position, velocity, sp ...
```
```suggestion
MCViNE simulations are run from a script that defines an instrument, where an instrument is composed of multiple components. A simple 4 component instrument is shown in Figure~\ref{fig:instrument}. Full instruments may have several hundred components. Each instrument script is run with a specified number of probability packets, where each packet has several state variables: position, velocity, spin, probability, and time.
```

Properly capitalize.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+MCViNE simulations are run from a script that defines an instrument, where an instrument is composed of multiple components. A simple 4 component instrument is shown in figure \ref{fig:instrument}. Full instruments may have several hundred components. Each instrument script is run with a specified number of probability packets, where each packet has several state variables: position, velocity, sp ...
+
+At a high-level, a component takes a neutron as input and performs some action on the neutron. Components can be attached to the instrument at a specified position and orientation. Some of the main component types are sources, guides, monitors, samples, and detectors. In a full instrument simulation, neutrons are generated from a source component and are then propagated through each component in  ...
```
```suggestion
At a high-level, a component takes a neutron as input and performs some action on the neutron. Components can be attached to the instrument at a specified position and orientation. Some of the main component types are sources, guides, monitors, samples, and detectors. In a full instrument simulation, neutrons are generated from a source component and are propagated through each component in the instrument. Sample components are a special type of component with additional input files to specify geometry and material composition.
```

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+\subsection{Component Hierarchy}
+
+One of the major benefits of using Python for this application is the ease of utilizing an object-oriented design and polymorphism.
+Each component inherits from a base \texttt{AbstractComponent} class. This \texttt{AbstractComponent} class requires a ``propagate'' method to be defined which takes in a neutron for the first parameter. The ``propagate'' method is responsible for defining the action of the component and is decorated with \texttt{@cudajit} to indicate that it is a CUDA kernel. An example of creating a simple comp ...
```
"can be seen below" Please do not make reference to direction in papers. Please reference the material specifically by its Figure/Equation/Table/Listing number.

matthewfeickert commented on papers/coleman_kendrick/code/arm.tex:
```diff
@@ -0,0 +1,23 @@
+\begin{Verbatim}[commandchars=\\\{\}]
+\PY{k+kn}{from} \PY{n+nn}{..ComponentBase} \PY{k+kn}{import} \PY{n}{ComponentBase}
```
Do not use a relative import in your example code. Code snippets are meant to be something a user would do, and so should be given as absolute imports from a distributed package.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+\subsection{CUDA Kernel Generation for an Instrument}
+
+In order to run a simulation from an input instrument script, \texttt{mcvine.acc} first generates a GPU kernel from the instrument specification. The input script will be parsed and then ``compiled'' to form an instrument kernel. This compiled version is then imported and executed to run the simulation. An example of a simple MCViNE instrument script containing a source, guide, and monitor can be ...
```
Why is "compiled" in quotes? If you view this as not actual compilation then please use more precise language until you have defined what you mean.

matthewfeickert replied:
Same comment as before on "seen below".

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+In order to run a simulation from an input instrument script, \texttt{mcvine.acc} first generates a GPU kernel from the instrument specification. The input script will be parsed and then ``compiled'' to form an instrument kernel. This compiled version is then imported and executed to run the simulation. An example of a simple MCViNE instrument script containing a source, guide, and monitor can be ...
+
+\vspace{1mm}
+\input{code/acc_sgm_instrument_trimmed.tex}
```
This is split across multiple columns. It would be better to reformat things, but if it is too painful/difficult this is okay as is.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+When \texttt{mcvine.acc} is called on the above instrument script, a new Python file is generated which contains a Numba CUDA kernel for the entire instrument. Each of the instrument components' \texttt{process} kernels are collected and inserted to form a generic kernel in this process. This generated kernel effectively models a neutron travelling through the entire instrument.
+
+Depending on the kernel launch configuration, each GPU thread might be responsible for more than one neutron. An example of a compiled instrument script can be seen in the code listing below. As seen in the script, a CUDA kernel is defined using Numba. Inside the kernel, each GPU thread will loop over the number of neutrons it is processing. Each \texttt{propagate} function has a number appended  ...
```
Same comment as before on "below".

matthewfeickert replied:
Your choice, but ``\texttt{propagate0}'' doesn't really need the quotes. Just having these as `texttt` would be fine.

matthewfeickert commented on papers/coleman_kendrick/code/acc_sgm_instrument_compiled.tex:
I don't think that this needs to be written like this to get the code formatting to work properly, but regardless, in here somewhere are lines that do single letter variable assignment. Please don't do this in code examples.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+One unique feature of MCViNE is its sample component, which allows for simulation of complex sample/sample environment and detector systems with flexible, sophisticated geometry and scattering physics. 
+This feature has enabled simulations that produce virtual experiment data that closely resemble real experimental data and make MCViNE a useful tool for both instrument design and advanced data analysis. 
+
+A Sample is made up of a Sample Assembly which has ``Shape'' and ``Phase'' tags. 
+The ``Shape'' tag contains the geometry specification and the ``Phase'' tag together with additional xml file(s) named ``\{name\}-scatterer.xml'' contain the scattering physics specification, where ``\{name\}'' is the name of a scatterer included in the sample assembly. A simple example for an aluminum (Al) sphere is shown below. 
```
It would be easier to read this if the terms here were typeset with `texttt` instead of quotes.

matthewfeickert replied:
Same comment as before RE: "shown below".

matthewfeickert replied:
```suggestion
The ``Shape'' tag contains the geometry specification and the ``Phase'' tag together with additional XML file(s) named ``\{name\}-scatterer.xml'' contain the scattering physics specification, where ``\{name\}'' is the name of a scatterer included in the sample assembly. A simple example for an aluminum (Al) sphere is shown below. 
```

XML is caps.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+Simple shapes can be created easily, but more complex shapes can be created too.
+To create the sample geometry, MCViNE uses Constructive Solid Geometry (CSG). 
+CSG can create complex geometries from operations such as intersection and union, on primitive shapes, such as cubes, spheres, and cylinders. 
+For example, figure \ref{fig:csg} shows how CSG is used to create an annular sample can from cylindrical primitives. This can is similar to those used in experiment design\cite{PhysRevB.107.094511}.
```
```suggestion
For example, Figure~\ref{fig:csg} shows how CSG is used to create an annular sample can from cylindrical primitives. This can is similar to those used in experiment design\cite{PhysRevB.107.094511}.
```

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+Ray tracing routines are implemented as CUDA kernels for these primitive shapes and operations. To support complex geometries that might involve many operations and shapes, the visitor pattern is used which constructs a single CUDA kernel to handle the ray intersection of that shape. This highlights one of the major advantages of using Python and Numba, as the ability to generate a CUDA kernel dy ...
+
+The specification of the scattering physics of a particular neutron scatterer is described in a dedicated "scatterer" xml file, where one or more sample kernels can be specified.
+An example of a scatterer xml file for specifying scattering physics can be seen below.
```
```suggestion
The specification of the scattering physics of a particular neutron scatterer is described in a dedicated "scatterer" XML file, where one or more sample kernels can be specified.
An example of a scatterer XML file for specifying scattering physics can be seen below.
```

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+\section{Results}
+
+Two types of comparisons were performed to show the usefulness of \texttt{mcvine.acc}. First, simulations comparing the CPU and GPU shows a significant performance gain by using a GPU.
```
Where do they show this? If it is in Figure 3 then reference it.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+\begin{figure}[h]
+    \centering
+    \includegraphics[width=\columnwidth]{figures/mcvine_hss_isotropic_timings.png}
```
See general review remark RE: figure axis labels.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+    \label{fig:results_ms}
+\end{figure}
+
+For the UN single scattering case, figure \ref{fig:results_ss} shows that for $10^{11}$ neutrons, the GPU version obtained a speedup of 383x over a single core, and 33x over 16 cores. For the UN multiple scattering case, figure \ref{fig:results_ms} shows that for $10^{10}$ neutrons, the GPU version obtained a speedup of 137x over a single core, and 10x over 16 cores. Comparing the speedup achieve ...
```
```suggestion
For the UN single scattering case, Figure~\ref{fig:results_ss} shows that for $10^{11}$ neutrons, the GPU version obtained a speedup of 383x over a single core, and 33x over 16 cores. For the UN multiple scattering case, Figure~\ref{fig:results_ms} shows that for $10^{10}$ neutrons, the GPU version obtained a speedup of 137x over a single core, and 10x over 16 cores. Comparing the speedup achieved for the simple isotropic sphere to the UN with single and multiple scattering shows the additional complexity required for the UN sample. A speedup of 10x over 16-core CPU for the UN multiple scattering case is still significant as some simulations can take on the order of days to months to complete.
```

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+\begin{figure}[h]
+\centering
+\includegraphics[width=\columnwidth]{figures/UN-ARCS-iqe.pdf}
```
As a color map is used here there should be information in the caption and ideally on the figure as well on how to interpret the color.

matthewfeickert replied:
@ckendrick Can you update the caption to explain the color map (a third dimension for this figure)?

ckendrick replied:
Sure, I will update the caption! Sorry, that must've slipped through.

matthewfeickert replied:
> Color indicates the scattering intensity where brighter regions represent higher intensity.

This is good enough, but out of curiosity are the actual intensity values not relevant, but more that things are just above threshold?

ckendrick replied:
I'm probably not the best person to answer this question but roughly yes, I believe that is correct.

The intensity values are produced during the data reduction step, and I believe this reduction step effectively normalizes the intensity based on the data.
For the experimental data, the intensity is essentially the total neutron event counts measured by the detectors in the instrument. Since the neutron event counts can depend on several factors (time of experiment, etc), I think it is more useful to look at the "normalized" intensity since it better captures the overall scattering response of the material.

For the simulation data, we determine the intensity by calculating the sum of all the probabilities of the Monte Carlo packets ("neutrons") that hit a particular area (for this case, these areas correspond to the actual instrument detectors since we are modeling the physical instrument geometry with the same material properties).

Hopefully that helps! If you're still interested, [the original paper with these results](http://dx.doi.org/10.1103/PhysRevB.89.144302) goes into more details of both the experimental setup and simulation.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+
+\begin{figure}[H]
+\centering
+\includegraphics[width=\columnwidth]{figures/Constant_E_cuts.pdf}
```
Please do not obscure visualization with plot legends. If something is going to be plotted it should be visible. Please resize the plot axes so that the legend is visible without obscuring any components of the plot.

matthewfeickert commented on papers/coleman_kendrick/mcvine.tex:
```diff
@@ -0,0 +1,193 @@
+\section{Acknowledgements}
+For initial Framework development and instrument component development, this research used resources of the Spallation Neutron Source Second Target Station Project at Oak Ridge National Laboratory (ORNL). Sample development was sponsored by ORNL's Laboratory Director's Research and Development Fund. ORNL is managed by UT-Battelle LLC for DOE’s Office of Science, under Contract No. DE-AC05-00OR227 ...
+
+The US government retains and the publisher, by accepting the article for publication, acknowledges that the US government retains a nonexclusive, paid-up, irrevocable, worldwide license to publish or reproduce the published form of this manuscript, or allow others to do so, for US government purposes. DOE will provide public access to these results of federally sponsored research in accordance w ...
```
Is there are citeable form of the DOE Public Access Plan?

matthewfeickert replied:
> Citable DOE public access plan: To my knowledge, I am not aware of a citable version of this. I can continue to look into this. However, we are required to display this full text and link even if there is a citable form available.

Okay. While it is always nice to have stable documents for citation if you're required to provide the full text and link then that's fine.

ckendrick replied:
I agree. I believe the link should hopefully be stable over time and will always direct to the page with the most current DOE Public Access Plan. You can see there is a 2014 version and a 2023 version currently at the link.
 
The other best citable form I found was the specific US regulation for this, but I'm not sure how stable this would be either (long-term):
https://www.ecfr.gov/current/title-48/part-970/section-970.5227-2

matthewfeickert replied:
> I believe the link should hopefully be stable over time and will always direct to the page with the most current DOE Public Access Plan. You can see there is a 2014 version and a 2023 version currently at the link.

Ah okay. :+1:

matthewfeickert reviewed (changes requested):
Thanks for the proceedings submission @ckendrick!

Overall this paper looks good but there are a few formatting errors that will need revision (an excellent example of a well formatted paper that deals with complex figures and image placing is [signac: Data Management and Workflows for Computational Researchers](https://conference.scipy.org/proceedings/scipy2021/bradley_dice.html) from SciPy 2021). Please see the comments I left as code review as well as the following high level review remarks. Please let me know if you have any questions as I am more than happy to elaborate (this review is perhaps more brief than intended, but is getting written while in transit).

This review follows the [Suggested Review Criteria for SciPy Proceedings](https://github.com/scipy-conference/scipy_proceedings/blob/620d77e44fc92fcead50479ce9b0eacc5825440a/review_criteria.md)

## Written Quality

- [x] Prose should be written in English.
- [x] Prose should clearly and consistently communicate the narrative (e.g. consistent tense and person; use of active voice; etc)
- [x] The written style should convey information that furthers the knowledge or research of the reader.
- [ ] Due to the interdisciplinary nature of SciPy, highly domain-specific jargon should be avoided or explained where possible.
   - There are a few things already brought up in the other review that could be improved.

## Technical Content

- [x] The technical content should be scientifically sound.
- [x] Computational content should, likewise, be verified.
- [x] The work should describe the development or use of python software for approaching a problem in a domain within the scope of the conference.

## Novelty and Impact

- [x] Assume that the novelty of the conceptual work described in the paper has been assessed by the Program Committee.
  The paper should advance the state of a scientific domain, the practice of scientific computing itself, or another subject area within the scope of the conference.

## Verifiability

- [x] Software descriptions should be accompanied by references to or examples of representative source code.
- [ ] Source code essential to the conclusions of the paper should be made available to the reader.
   - Is there a repository of the code used to produce the plots?
- [x] Data sources should be identified (e.g., with citation to a persistent DOI).
   - No public datasets are used
- [ ] Analysis should be accompanied by a workflow description sufficient to reproduce the results.
   - Is there a repository of the code used to produce the plots?

## Other Requirements

- [ ] All mentioned software should be formally and consistently cited wherever possible.
   - DOIs are missing for most citations.
- [x] Acronyms should be spelled out upon first mention.
- [x] License conditions on images and figures must be respected (Creative Commons, etc.).
- [x] Mathematical and other symbols should be defined.
- [x] Definitions should include consistent units where appropriate.
- [ ] Avoid long paragraphs of your rich text file to enable reviewers to target specific sections with their comments. GitHub comments are per line of text.
   - One sentence per line here would have made comments much easier. Please follow this in future submissions.
- [x] Avoid custom LaTeX markup where possible.

### Length

- [x] The compiled version should be no longer than 8 pages, including figures, source code, appendices. References are excluded from this limit.

### Figures

- [x] All figures and tables should have captions.
   - Code snippet text should also be cpationed
- [ ] Figure text should be of a readable size.
   - Figures axis labels are too small and should not be in all lowercase.
- [x] Images and figures should be reasonably sized and formatted for viewing online; typically a few hundred kilobytes and less than 1 MB.

### Code Snippets

- [ ] Code snippets should be formatted to fit inside a single column without overflow. Please make sure to check the build server's copy of your paper; it's the authoritative view of how your paper will appear.
     - Code snippet 2 overflows, which is not great, but liveable.
- [x] If not in Python, the language in the code snippet should be mentioned.
   - If code snippets had captions this could be achieved.
- [x] Code snippets should follow a common style guide. PEP8 is preferred.
   - Good enough I think. Everything is readable even if it doesn't follow a standard.

mepa commented:
@scoobies mark pending comment

deniederhut commented:
Hey @ckendrick ! Have you had a chance to revise your paper based on the reviewer comments?

ckendrick committed:
Update paper from reviews

ckendrick committed:
Fix typo in table

ckendrick commented:
@sanhitamj @matthewfeickert Thank you both very much for your comments and taking the time to review the paper! I believe we have addressed most of the comments you had. There were a few instances where we didn't which I will elaborate on below, but most of these I believe are technical limitations rather than disagreement with your comments.

@sanhitamj I believe we addressed everything except for the full form of LLVM. LLVM is not an acronym, but the name of the project itself (see the first paragraph of the overview on their website: https://llvm.org/).

@matthewfeickert I believe the only things we did not address are listed below:

- Code listings: I agree that these should be numbered and labeled so that we can refer to them in-text rather than saying "see below..". However, I could not find an easy way to do that with the current SciPy latex style. Unfortunately, we have to wrap everything in a `verbatim` environment, and I could not find an option in the `fancyvrb` that allowed this (this is also why the code listings are done the way they are - we had to use pygmentize in order to generate Python code coloring).
Normally we would use packages like `listings` or `minted` which have support for formatting, numbered listings, captions, and labels, but I believe we cannot include additional packages with the current SciPy latex style. I looked through several other SciPy proceedings paper, and it seems this is a common issue. Several other papers have to refer to listings in-text based on direction rather than a label. Please let me know if there is a way around this that I missed, or if this would be something we should ask an editor about?
- Single variable assignment in the generated SGM listing: I agree that this is not ideal, though this code is not necessarily meant for users to ever see or interact with. This particular code gets generated behind-the-scenes from the user, we just wanted to show it in the paper to help explain what we are doing. We can change the variables to have more accurate meaning (`v` = `velocity`, `r` = `position`) but this might also lead to more lines exceeding the length of the text column and having to be split over other lines. We definitely don't mind changing this if you think this would be clearer.
- Multiple citation style: We agree that having a large list of citations in-text looks a bit funny. Since that is part of the IEEE style used, we left it as is. We can change this to using a list if you think this would be better, such as [3]-[8]. It was unclear to me if this would be acceptable to do or not.
- Citable DOE public access plan: To my knowledge, I am not aware of a citable version of this. I can continue to look into this. However, we are required to display this full text and link even if there is a citable form available.
- RE code used to generate figures: The Github repo that was originally linked in the paper (and now cited as a reference) was used to generate the figures (data displayed for the UN results, + CPU/GPU timings via a script that automatically generates this). Is this something we should mention explicitly in the paper?
- RE citation DOIs: Thanks for pointing this out! I believe these should all be fixed, the DOI checker probably needs to be re-run. If there is any missing still, we can certainly fix those as well.

Thanks again for all of your comments! Let me know if we missed something and we will fix it.

mepa commented:
Thanks, @ckendrick!

@sanhitamj, @matthewfeickert, how are you feeling about the updates made by the author?

matthewfeickert commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1016/j.physb.2018.11.042 is OK
- 10.1063/1.5007089 is OK
- 10.1063/1.5116147 is OK
- 10.1063/5.0079031 is OK
- 10.1103/physrevb.105.l180403 is OK
- 10.1063/5.0086451 is OK
- 10.1063/5.0089783 is OK
- 10.1063/5.0089740 is OK
- 10.1063/5.0090919 is OK
- 10.1063/5.0090810 is OK
- 10.1063/5.0089524 is OK
- 10.1063/5.0090527 is OK
- 10.1063/5.0089349 is OK
- 10.1063/5.0122279 is OK
- 10.1016/j.nima.2022.167881 is OK
- 10.3233/JNR-190108 is OK
- 10.3233/JNR-200186 is OK
- 10.1063/1.4870050 is OK
- 10.1088/1742-6596/251/1/012058 is OK
- 10.1038/ncomms2117 is OK
- 10.1016/j.nima.2015.11.118 is OK
- 10.1103/physrevb.89.144302 is OK
- 10.1103/PhysRevB.107.094511 is OK
- 10.13139/ORNLNCCS/1975747 is OK

MISSING DOIs

- 10.1063/1.3680104 may be a valid DOI for title: Design and operation of the wide angular-range chopper spectrometer ARCS at the Spallation Neutron Source
- 10.1016/j.nima.2015.11.118 may be a valid DOI for title: MCViNE–An object oriented Monte Carlo neutron ray tracing simulation package

INVALID DOIs

- https://doi.org/10.1016/j.cpc.2017.04.012 is INVALID because of 'https://doi.org/' prefix
```

matthewfeickert commented on papers/coleman_kendrick/bib.bib:
```diff
@@ -0,0 +1,431 @@
+Program summary
+Program Title: MCPL Program Files doi: http://dx.doi.org/10.17632/cby92vsv5g.1 Licensing provisions: CC0 for core MCPL, see LICENSE file for details. Programming language: C and C++ External routines/libraries: Geant4, MCNP, McStas, McXtrace Nature of problem: Saving particle states in Monte Carlo simulations, for interchange between simulation packages or for reuse within a single package. Solut ...
+	author = {T. Kittelmann and E. Klinkby and E.B. Knudsen and P. Willendrup and X.X. Cai and K. Kanaki},
+	doi = {https://doi.org/10.1016/j.cpc.2017.04.012},
```
```suggestion
	doi = {10.1016/j.cpc.2017.04.012},
```

Fixes the

```
INVALID DOIs

- https://doi.org/10.1016/j.cpc.2017.04.012 is INVALID because of 'https://doi.org/' prefix
```

error

matthewfeickert commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper ckendrick-2023_mcvine scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

matthewfeickert reviewed (changes requested):
Thanks for your revisions and comments @ckendrick (also you gave a very nice talk at SciPy too!)

> * Code listings: I agree that these should be numbered and labeled so that we can refer to them in-text rather than saying "see below..". However, I could not find an easy way to do that with the current SciPy latex style. Unfortunately, we have to wrap everything in a `verbatim` environment, and I could not find an option in the `fancyvrb` that allowed this (this is also why the code listings are done the way they are - we had to use pygmentize in order to generate Python code coloring).
>   Normally we would use packages like `listings` or `minted` which have support for formatting, numbered listings, captions, and labels, but I believe we cannot include additional packages with the current SciPy latex style. I looked through several other SciPy proceedings paper, and it seems this is a common issue. Several other papers have to refer to listings in-text based on direction rather than a label. Please let me know if there is a way around this that I missed, or if this would be something we should ask an editor about?

@mepa do you have any advice on this? If no, then we can just file this as a thing to make easier next year.

> ... We can change the variables to have more accurate meaning (`v` = `velocity`, `r` = `position`) but this might also lead to more lines exceeding the length of the text column and having to be split over other lines. We definitely don't mind changing this if you think this would be clearer.

@ckendrick I'll leave it up to you. I in general try to wage war on single letter variables in code anywhere, but I appreciate that you're also trying to get code into a 2-column paper and sacrifices need to be made for any of that to work. :) 

---

Overall this is good! Beyond the DOI fix that I suggested if you can clarify the image contents of Figure 6 in the caption then I think this is good to go. :+1:

ckendrick committed:
Fix some additional typos

ckendrick committed:
Use ~\ref for figures to help prevent line breaks

ckendrick committed:
Add missing spaces before \cite

ckendrick committed:
Re-fix invalid DOI after overwriting bib file

ckendrick committed:
Fix caps, middle names, and add more dois in bib

matthewfeickert commented:
@scoobies check references

scoobies commented:
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1063/1.3680104 is OK
- 10.1145/2833157.2833162 is OK
- 10.1016/j.physb.2018.11.042 is OK
- 10.1063/1.5007089 is OK
- 10.1063/1.5116147 is OK
- 10.1063/5.0079031 is OK
- 10.1103/PhysRevX.9.011035 is OK
- 10.1103/PhysRevB.105.214405 is OK
- 10.1103/physrevb.105.l180403 is OK
- 10.1063/5.0086451 is OK
- 10.1063/5.0089783 is OK
- 10.1063/5.0089740 is OK
- 10.1063/5.0090919 is OK
- 10.1063/5.0090810 is OK
- 10.1063/5.0089524 is OK
- 10.1063/5.0090527 is OK
- 10.1063/5.0089349 is OK
- 10.1063/5.0122279 is OK
- 10.1063/5.0131778 is OK
- 10.1016/j.nima.2022.167881 is OK
- 10.3233/JNR-190108 is OK
- 10.3233/JNR-200186 is OK
- 10.1063/1.4870050 is OK
- 10.1088/1742-6596/251/1/012058 is OK
- 10.1038/ncomms2117 is OK
- 10.1016/j.physb.2006.05.281 is OK
- 10.1088/2399-6528/ab3622 is OK
- 10.1016/j.nima.2015.11.118 is OK
- 10.1103/physrevb.89.144302 is OK
- 10.1103/PhysRevB.107.094511 is OK
- 10.13139/ORNLNCCS/1975747 is OK
- 10.1016/j.cpc.2017.04.012 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

ckendrick committed:
Update caption to include color map description

ckendrick commented:
@matthewfeickert Thank you! I appreciate you going over the changes again. It looks like the DOI issues are finally fixed. Let me know if there is anything else that needs to be changed.

ckendrick commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper ckendrick-2023_mcvine scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

mepa commented:
@scoobies mark ready

deniederhut reviewed (approved):
Thanks @matthewfeickert and @sanhitamj for your help getting this paper ready for publication! And thanks @ckendrick for joining us at SciPy this year!!

Did I say thanks yet? SciPy turns me into a thank-you machine.

![giphy-3908445652](https://github.com/scipy-conference/scipy_proceedings/assets/6868396/16c270b8-3c08-4723-a7c0-f587741469d3)

matthewfeickert commented:
Thank you, @deniederhut, for your continued leadership and heroic work in leading and facilitating the proceedings for yet another year! :)

deniederhut commented:
🙇
